// components/Feed.js
import React from 'react'
import Episode from './Episode'


export default class Feed extends React.Component {
    constructor() {
        super();
        this.state = {
            pattern: ""
        }
    }

    getEpisodes() {
        this.props.addEpisodes(this.props.episodes.length+1);
    }

    wow() {
        const pattern = this.refs.pattern.value.trim().toLowerCase();
        this.setState({
            pattern: pattern
        });
    }

    render() {
        const { episodes, isLoading, error }  = this.props;
        console.log(episodes[0]);
        const { pattern } = this.state;

        const wow = episodes.filter(episode => episode.serial.title.toLowerCase().indexOf(pattern) + 1);

        return (
            <div className="row">
                <input type="text" ref="pattern" id="pattern" onChange={() => this.wow()} />
                <div className="row">
                    <div className="large-6 large-centered columns">
                        {wow.map(function (episode) {
                            return <Episode key={episode.id} episode={episode}/>
                        })}
                    </div>
                </div>
                <div className="row">
                    <div className="large-7 large-centered columns">
                        <a href="#" onClick={() => this.getEpisodes()} className="button expand">Expanded Button</a>
                    </div>
                </div>

                {() => {
                    if (isLoading == true) {
                        return (
                            <div className="row">Загрузка...</div>
                        )
                    }
                    if (error) {
                        return (
                            <div className="row">{error}</div>
                        )
                    }
                }}
            </div>
        )
    }

}
